crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -279,66 +295,343 @@ where
         Ok(())
     }
 
+    /// Look up the IBC denomination from a IbcToken.
+    pub fn query_ibc_denom(
+        &self,
+        token: impl AsRef<str>,
+        owner: Option<&Address>,
+    ) -> Result<String> {
+        let hash = match Address::decode(token.as_ref()) {
+            Ok(Address::Internal(InternalAddress::IbcToken(hash))) => {
+                hash.to_string()
+            }
+            _ => return Ok(token.as_ref().to_string()),
+        };
+
+        if let Some(owner) = owner {
+            let ibc_trace_key = ibc_trace_key(owner.to_string(), &hash);
+            if let Some(ibc_denom) = self.ctx.read_pre(&ibc_trace_key)? {
+                return Ok(ibc_denom);
+            }
+        }
+
+        // No owner is specified or the owner doesn't have the token
+        let ibc_denom_prefix = ibc_trace_key_prefix(None);
+        let ibc_denoms = self.ctx.iter_prefix(&ibc_denom_prefix)?;
+        for (key, ibc_denom, _gas) in ibc_denoms {
+            if let Some((_, token_hash)) =
+                is_ibc_trace_key(&Key::parse(key).into_storage_result()?)
+            {
+                if token_hash == hash {
+                    return String::try_from_slice(&ibc_denom[..])
+                        .into_storage_result()
+                        .map_err(Error::NativeVpError);
+                }
+            }
+        }
+
+        Ok(token.as_ref().to_string())
+    }
+
+    // Apply the given send packet to the changed balances structure
+    fn apply_send_packet(
+        &self,
+        acc: &mut ChangedBalances,
+        msg: &MsgTransfer,
+        keys_changed: &BTreeSet<Key>,
+    ) -> Result<()> {
+        // Compute the packet commitment for this message
+        let Ok(packet_data_bytes) =
+            serde_json::to_vec(&msg.message.packet_data)
+        else {
```

Comment:
When would this case happen? Seems concerning, no?

Reply:
Assuming the correct implementation of `Serialize` on `PacketData`, I can't think of anything that would cause this case to happen. The point of this line had merely been that in this case `Tx` validation could theoretically correctly continue without adding this Packet to `ChangedBalances`. That being said, propagating an error is also a possibility. Would you prefer this or something else?

Reply:
I think it's safer to reject the transaction in these sorts of cases - just defense-in-depth.